Rewrite docs for `std::ptr`

ecstatic-morse opened the pull request:
This PR attempts to resolve #29371.

This is a fairly major rewrite of the `std::ptr` docs, and deserves a fair bit of scrutiny. It adds links to the GNU libc docs for various instrinsics, adds internal links to types and functions referenced in the docs, adds new, more complex examples for many functions, and introduces a common template for discussing unsafety of functions in `std::ptr`.

All functions in `std::ptr` (with the exception of `ptr::eq`) are unsafe because they either read from or write to a raw pointer. The "Safety" section now informs that the function is unsafe because it dereferences a raw pointer and requires that any pointer to be read by the function points to "a valid vaue of type `T`".

Additionally, each function imposes some subset of the following conditions on its arguments.

* The pointer points to valid memory.
* The pointer points to initialized memory.
* The pointer is properly aligned.

These requirements are discussed in the "Undefined Behavior" section along with the  consequences of using functions that perform bitwise copies without requiring `T: Copy`. I don't love my new descriptions of the consequences of making such copies. Perhaps the old ones were good enough?

Some issues which still need to be addressed before this is merged:
- [ ] The new docs assert that `drop_in_place` is equivalent to calling `read` and discarding the value. Is this correct?
- [ ] Do `write_bytes` and `swap_nonoverlapping` require properly aligned pointers?
- [ ] The new example for `drop_in_place` is a lackluster.
- [ ] Should these docs rigorously define what `valid` memory is? Or should is that the job of the reference? Should we link to the reference?
- [ ] Is it correct to require that pointers that will be read from refer to "valid values of type `T`"?
- [x] I can't imagine ever using `{read,write}_volatile` with non-`Copy` types.  Should I just link to {read,write} and say that the same issues with non-`Copy` types apply?
- [x] `write_volatile` doesn't link back to `read_volatile`.
- [ ] Update docs for the unstable [`swap_nonoverlapping`](https://github.com/rust-lang/rust/issues/42818)
- [ ] Update docs for the unstable [unsafe pointer methods RFC](https://github.com/rust-lang/rfcs/pull/1966)

Looking forward to your feedback.

r? @steveklabnik

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @steveklabnik (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. Due to the way GitHub handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [the contribution instructions](https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md) for more information.

TimNN commented:
Your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/363611878). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
[00:00:46] configure: rust.quiet-tests     := True
---
[00:05:21] tidy error: /checkout/src/libcore/intrinsics.rs:995: line longer than 100 chars
[00:05:21] tidy error: /checkout/src/libcore/intrinsics.rs:1152: unexplained "```ignore" doctest; try one:
[00:05:21]
[00:05:21] * make the test actually pass, by adding necessary imports and declarations, or
[00:05:21] * use "```text", if the code is not Rust code, or
[00:05:21] * use "```compile_fail,Ennnn", if the code is expected to fail at compile time, or
[00:05:21] * use "```should_panic", if the code is expected to fail at run time, or
[00:05:21] * use "```no_run", if the code should type-check but not necessary linkable/runnable, or
[00:05:21] * explain it like "```ignore (cannot-test-this-because-xxxx)", if the annotation cannot be avoided.
```

</details><p></p>

I'm a bot! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**.

ecstatic-morse commented:
The CI build failed because I added an ignored doctest to `write_bytes` which demonstrates how `write_bytes` can still cause undefined behavior even if the memory to be written is valid. It does this by overwriting a `Box<T>` with all zeroes. This example is fairly trivial; I could remove it, but is there a better way to fix this?

TimNN commented:
Your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/363633803). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
[00:00:50] configure: rust.quiet-tests     := True
---
[00:42:33] ............................................................................i.......................
[00:42:39] ...................i................................................................................
---
[00:43:14] ................................................................................................i...
[00:43:21] ......................................................................i.............................
---
[00:44:14] .............................................i......................................................
---
[00:48:02] .............................i......................................................................
[00:48:16] ..............................................................i.....................................
[00:48:32] ...............................................i....................................................
[00:48:51] ....................................................................................................
[00:49:12] ....................................................................................................
[00:49:33] ....................................................................................................
[00:49:58] ...i...............................................................................................i
[00:50:26] ......................................................................................test [run-pass] run-pass/mir_heavy_promoted.rs has been running for over 60 seconds
[00:50:33] ..............
[00:51:03] ....................................................................................................
[00:51:36] .................................................................ii.................................
[00:52:24] ............................i....................................................i.ii..........test [run-pass] run-pass/saturating-float-casts.rs has been running for over 60 seconds
[00:52:27] .....
[00:53:06] .........................................................................................iiiiiii....
---
[00:55:17] ...................i............................................................ii.iii..............
[00:55:25] ....................................................................................................
[00:55:32] .........i..............................i...........................................................
[00:55:40] ....................................................................................................
[00:55:46] .....................i..............................................................................
[00:55:54] ....................................................................................................
[00:56:04] ....................................................................................................
[00:56:14] ....................................................................................................
[00:56:25] ....................................................................................................
[00:56:37] ....................................................................................................
[00:56:45] ..............i.....................................................................................
[00:56:54] ..................i..ii.............................................................................
[00:57:04] ....................................................................................................
[00:57:14] ....................................................................................................
[00:57:23] .....................................................................................i..............
[00:57:34] ...............................i....................................................................
---
[00:58:10] ...........................i........................................................................
[00:58:11] ....................................................................i...............................
[00:58:12] ................i.......................................................
---
[00:58:27] .............i........................
---
[00:58:56] i...i..ii....i.............ii.........iii......i..i...i...ii..i..i..ii.....
---
[00:59:00] i.......i......................i.......
---
[00:59:38] iiii.......i..i........i..i.i.............i..........iiii...........i...i..........ii.i.i.......ii..
[00:59:39] ....ii...
---
[01:08:08] .....i..............................................................................................
---
[01:09:56] .....................................i..............................................................
[01:10:15] ....................................................................................................
[01:10:34] ..........................................i.........................................................
---
[01:11:59] .........................................................ii.........................................
---
[01:13:00] ...............................................................i....................................
---
[01:17:51] ii..................................................................................................
[01:18:10] ....................................................................................................
[01:18:25] ...................iii......i......i...i......i.....................................................
[01:18:34] ....................................................................................................
[01:18:50] .........................................iiii........ii.............................................
[01:19:00] ....................................................................................................
[01:19:17] ........................................................................................i...........
[01:19:39] ....................................................................................................
[01:19:50] ..................................................................................................ii
[01:19:59] ii...............................................
---
[01:26:09] std/intrinsics/fn.drop_in_place.html:64: broken link - std/intrinsics/fn.read.html
[01:26:09] std/intrinsics/fn.drop_in_place.html:73: broken link - std/intrinsics/fn.read.html
[01:26:09] std/intrinsics/fn.drop_in_place.html:94: broken link - std/intrinsics/fn.write.html
[01:26:09] std/intrinsics/fn.copy_nonoverlapping.html:92: broken link - std/marker/trait.Copy
[01:26:09] std/ptr/fn.copy_nonoverlapping.html:92: broken link - std/marker/trait.Copy
[01:26:16] core/intrinsics/fn.copy_nonoverlapping.html:92: broken link - core/marker/trait.Copy
---
121636 ./obj/build/bootstrap/debug/incremental/bootstrap-351vorei3hhuv/s-ezws74tyww-1fcsc5e-1a7xax1ixldkk
---
102728 ./obj/build/x86_64-unknown-linux-gnu/stage1-std/x86_64-unknown-linux-gnu/release/incremental/core-31lccp6wy7orz/s-ezwt7v3hh8-khfl1c-2jdqxinbh81q7
100656 ./obj/build/bootstrap/debug/incremental/bootstrap-zki88qpb4pjm
100652 ./obj/build/bootstrap/debug/incremental/bootstrap-zki88qpb4pjm/s-ezwta9aetu-fj70km-2morwobbft3l8
```

</details><p></p>

I'm a bot! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**.

TimNN commented:
Your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/363633262). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
[00:00:47] configure: rust.quiet-tests     := True
---
[00:48:00] ............................................................................i.......................
[00:48:07] ...................i................................................................................
---
[00:48:44] ................................................................................................i...
[00:48:51] ......................................................................i.............................
---
[00:49:49] .............................................i......................................................
---
[00:53:52] .............................i......................................................................
[00:54:07] ..............................................................i.....................................
[00:54:24] ...............................................i....................................................
[00:54:47] ....................................................................................................
[00:55:10] ....................................................................................................
[00:55:32] ....................................................................................................
[00:55:57] ...i...............................................................................................i
[00:56:24] ..................................................................................test [run-pass] run-pass/mir_heavy_promoted.rs has been running for over 60 seconds
[00:56:35] ..................
[00:57:06] ....................................................................................................
[00:57:40] .................................................................ii.................................
[00:58:28] ............................i....................................................i.ii......test [run-pass] run-pass/saturating-float-casts.rs has been running for over 60 seconds
[00:58:33] .........
[00:59:14] .........................................................................................iiiiiii....
---
[01:01:27] ...................i............................................................ii.iii..............
[01:01:35] ....................................................................................................
[01:01:43] .........i..............................i...........................................................
[01:01:50] ....................................................................................................
[01:01:57] .....................i..............................................................................
[01:02:05] ....................................................................................................
[01:02:15] ....................................................................................................
[01:02:26] ....................................................................................................
[01:02:37] ....................................................................................................
[01:02:50] ....................................................................................................
[01:02:59] ..............i.....................................................................................
[01:03:09] ..................i..ii.............................................................................
[01:03:20] ....................................................................................................
[01:03:32] ....................................................................................................
[01:03:41] .....................................................................................i..............
[01:03:53] ...............................i....................................................................
---
[01:04:32] ...........................i........................................................................
[01:04:34] ....................................................................i...............................
[01:04:35] ................i.......................................................
---
[01:04:52] .............i........................
---
[01:05:26] i...i..ii....i.............ii.........iii......i..i...i...ii..i..i..ii.....
---
[01:05:29] i.......i......................i.......
---
[01:06:10] iiii.......i..i........i..i.i.............i..........iiii...........i...i..........ii.i.i.......ii..
[01:06:12] ....ii...
---
[01:15:10] .....i..............................................................................................
---
[01:17:17] .....................................i..............................................................
[01:17:37] ....................................................................................................
[01:17:58] ..........................................i.........................................................
---
[01:19:40] .........................................................ii.........................................
---
[01:20:48] ...............................................................i....................................
---
[01:26:04] ii..................................................................................................
[01:26:25] ....................................................................................................
[01:26:43] ...................iii......i......i...i......i.....................................................
[01:26:53] ....................................................................................................
[01:27:11] .........................................iiii........ii.............................................
[01:27:22] ....................................................................................................
[01:27:40] ........................................................................................i...........
[01:28:04] ....................................................................................................
[01:28:15] ..................................................................................................ii
[01:28:25] ii...............................................
---
[01:34:59] std/intrinsics/fn.drop_in_place.html:64: broken link - std/intrinsics/fn.read.html
[01:34:59] std/intrinsics/fn.drop_in_place.html:73: broken link - std/intrinsics/fn.read.html
[01:34:59] std/intrinsics/fn.drop_in_place.html:94: broken link - std/intrinsics/fn.write.html
[01:34:59] std/intrinsics/fn.copy_nonoverlapping.html:92: broken link - std/marker/trait.Copy
[01:34:59] std/ptr/fn.copy_nonoverlapping.html:92: broken link - std/marker/trait.Copy
[01:35:06] core/intrinsics/fn.copy_nonoverlapping.html:92: broken link - core/marker/trait.Copy
[01:35:11] thread 'main' panicked at 'found some broken links', tools/linkchecker/main.rs:49:9
---
[01:35:11] make: *** [check] Error 1
[01:35:11] Makefile:58: recipe for target 'check' failed
```

</details><p></p>

I'm a bot! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**.

TimNN commented:
Your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/363634608). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
[00:00:47] configure: rust.quiet-tests     := True
---
[00:49:43] ............................................................................i.......................
[00:49:49] ...................i................................................................................
---
[00:50:27] ................................................................................................i...
[00:50:34] ......................................................................i.............................
---
[00:51:34] .............................................i......................................................
---
[00:55:44] .............................i......................................................................
[00:56:00] ..............................................................i.....................................
[00:56:16] ...............................................i....................................................
[00:56:38] ....................................................................................................
[00:57:01] ....................................................................................................
[00:57:24] ....................................................................................................
[00:57:50] ...i...............................................................................................i
[00:58:16] .......................................................................test [run-pass] run-pass/mir_heavy_promoted.rs has been running for over 60 seconds
[00:58:29] .............................
[00:59:03] ....................................................................................................
[00:59:41] .................................................................ii.................................
[01:00:26] ............................i............................................test [run-pass] run-pass/saturating-float-casts.rs has been running for over 60 seconds
[01:00:36] ........i.ii...............
[01:01:20] .........................................................................................iiiiiii....
---
[01:03:38] ...................i............................................................ii.iii..............
[01:03:46] ....................................................................................................
[01:03:54] .........i..............................i...........................................................
[01:04:02] ....................................................................................................
[01:04:09] .....................i..............................................................................
[01:04:18] ....................................................................................................
[01:04:28] ....................................................................................................
[01:04:39] ....................................................................................................
[01:04:50] ....................................................................................................
[01:05:04] ....................................................................................................
[01:05:13] ..............i.....................................................................................
[01:05:23] ..................i..ii.............................................................................
[01:05:33] ....................................................................................................
[01:05:44] ....................................................................................................
[01:05:54] .....................................................................................i..............
[01:06:05] ...............................i....................................................................
---
[01:06:48] ...........................i........................................................................
[01:06:50] ....................................................................i...............................
[01:06:52] ................i.......................................................
---
[01:07:08] .............i........................
---
[01:07:39] i...i..ii....i.............ii.........iii......i..i...i...ii..i..i..ii.....
---
[01:07:43] i.......i......................i.......
---
[01:08:23] iiii.......i..i........i..i.i.............i..........iiii...........i...i..........ii.i.i.......ii..
[01:08:24] ....ii...
---
[01:18:32] .....i..............................................................................................
---
[01:20:36] .....................................i..............................................................
[01:20:57] ....................................................................................................
[01:21:18] ..........................................i.........................................................
---
[01:22:56] .........................................................ii.........................................
---
[01:24:04] ...............................................................i....................................
---
[01:29:18] ii..................................................................................................
[01:29:38] ....................................................................................................
[01:29:54] ...................iii......i......i...i......i.....................................................
[01:30:04] ....................................................................................................
[01:30:20] .........................................iiii........ii.............................................
[01:30:31] ....................................................................................................
[01:30:48] ........................................................................................i...........
[01:31:14] ....................................................................................................
[01:31:25] ..................................................................................................ii
[01:31:34] ii...............................................
---
[01:38:28] std/intrinsics/fn.drop_in_place.html:64: broken link - std/intrinsics/fn.read.html
[01:38:28] std/intrinsics/fn.drop_in_place.html:73: broken link - std/intrinsics/fn.read.html
[01:38:28] std/intrinsics/fn.drop_in_place.html:94: broken link - std/intrinsics/fn.write.html
[01:38:28] std/intrinsics/fn.copy_nonoverlapping.html:92: broken link - std/marker/trait.Copy
[01:38:29] std/ptr/fn.copy_nonoverlapping.html:92: broken link - std/marker/trait.Copy
[01:38:36] core/intrinsics/fn.copy_nonoverlapping.html:92: broken link - core/marker/trait.Copy
---
$ ls -lat $HOME/Library/Logs/DiagnosticReports/
ls: cannot access /home/travis/Library/Logs/DiagnosticReports/: No such file or directory
```

</details><p></p>

I'm a bot! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**.

ExpHP commented on src/libcore/intrinsics.rs:
```diff
@@ -959,59 +959,131 @@ extern "rust-intrinsic" {
+    /// The caller must ensure that `src` points to a valid sequence of type
+    /// `T`.
+    ///
+    /// # Undefined Behavior
```
I don't see the role of having both `Safety` and `Undefined Behavior` sections. It seems to me that the convention is to have a single section named `Safety`.

```sh
[ lampam @ 21:52:55 ] (master •58 -7 +7) ~/asd/clone/rust
$ rg '# Undefined [bB]ehavior'
src/libcore/mem.rs
532:/// # Undefined behavior

src/liballoc/raw_vec.rs
154:    /// # Undefined Behavior
171:    /// # Undefined Behavior

$ rg '# Safety'
(far more results)
```

ExpHP replied:
(granted, I often wish the `# Safety` sections tended to more closely resemble your `# Undefined Behavior` sections, but that is just my opinion)

ecstatic-morse replied:
@steveklabnik [mentioned](https://github.com/rust-lang/rust/issues/29371#issuecomment-379234472) that new docs should make use of `Undefined Behavior` (I think upper case is correct). Yeah, the safety section is basically the same for every function. All these functions are trivially unsafe because they dereference raw pointers, and the `Safety` section as I have it doesn't add any new information. Perhaps it could be moved to the module docs and not repeated everywhere?

steveklabnik replied:
Originally, I felt like these two cases were pretty distinct, but now, re-thinking about it, I'm not so sure, honestly. In some senses, UB is a subset of safety, and it feels like it's worth calling out separately, but maybe not. dunno.

scottmcm replied:
I'm confused; I thought UB was the consequence of not following Safety.  Is there a kind of unsafety that doesn't lead to UB?

ecstatic-morse replied:
My thought process:

When attempting to divine what belonged under each heading while writing this, I took the point of view that `Safety` referred to the language feature (i.e. the subset of rust you must opt into with the `unsafe` keyword). Some functions (e.g. `Vec::set_len`) are not obviously unsafe from looking at their function signature. The `Safety` section is probably a good place to discuss this.

In discussing why a function is `unsafe`, one must communicate what invariants must be maintained to avoid UB. Most of the standard library uses the `Safety` heading to communicate this. I am now of the opinion that there's never really a reason to have both of these top-level headings (as I do now) since the information is so closely related.

However, for functions whose unsafety is obvious because they take a raw pointer and have names like `read`, `write`, and `copy`, it seemed clearer to name the section discussing its invariants `Undefined Behavior`. `RawVec::from_raw_parts` also uses this convention.

steveklabnik replied:
@scottmcm "safety" is "memory safety", there are kinds of UB that are not memory safety related.

ecstatic-morse commented on src/libcore/ptr.rs:
```diff
@@ -328,21 +507,33 @@ pub unsafe fn read_unaligned<T>(src: *const T) -> T {
+///
+/// # Undefined Behavior
+///
+/// `write` can trigger undefined behavior if any of the following conditions
```
Whoops, I agree, I thought I changed this for all functions to "Behavior is undefined if any of the following conditions are violated:", but managed to miss `write*`.

steveklabnik commented on src/libcore/ptr.rs:
```diff
@@ -10,7 +10,7 @@
 // FIXME: talk about offset, copy_memory, copy_nonoverlapping_memory
 
-//! Raw, unsafe pointers, `*const T`, and `*mut T`.
+//! Manually manage memory through raw, unsafe pointers.
```
I would drop the "unsafe" bit here and just use "through raw pointers"

pietroalbini commented:
Ping from triage! The reviewer asked a signoff from @rust-lang/libs, can someone check this PR?

pietroalbini commented:
Ping from triage @rust-lang/libs! Can someone check if the docs here are correct?

alexcrichton commented:
I'd personally prefer to fold the safety/undefined behavior sections together and also ensure it's worded something along the lines of "if violated these conditions will cause the function to be unsafe ..." instead of "this function is only unsafe if at least one of these conditions is violated ..." in the sense that these are descriptions of safety but not necessarily 100% exhaustive

TimNN commented:
Ping from triage, @ecstatic-morse ! @alexcrichton suggested some changes, do you think you'll have time to work on those?

ecstatic-morse commented:
@TimNN Sorry, I was busy studying for Code Jam this past week. I'll merge the sections after work today.

TimNN commented:
@ecstatic-morse: Awesome to hear, I hope the Code Jam went well! Please don't feel stressed / obligated to fix this ASAP by my comment. We're just regularly checking in with PRs to ensure stuff doesn't get forgotten.

ecstatic-morse commented:
@alexcrichton The current wording is "Behavior is undefined if any of the following conditions are violated:"

I think this communicates that the list is not necessarily exhaustive. Is this suitable?

I'm merging the sections now.

Gankro commented on src/libcore/intrinsics.rs:
```diff
@@ -1028,15 +1100,39 @@ extern "rust-intrinsic" {
+    /// # Safety
+    ///
+    /// `write_bytes` is unsafe because it dereferences a raw pointer. The
+    /// caller must ensure that the poiinter points to a valid value of type `T`.
```
typo: "poiinter"

Gankro commented on src/libcore/intrinsics.rs:
```diff
@@ -959,59 +959,131 @@ extern "rust-intrinsic" {
+    /// Behavior is undefined if any of the following conditions are violated:
+    ///
+    /// * The region of memory which begins at `src` and has a length of
+    ///   `count * size_of::<T>()` bytes must be *both* valid and initialized.
```
It is both unclear to me what "valid" means (a single live allocation..?). 

Also I don't understand why src must be initialized. Where did you get this from? This would make it impossible to implement realloc in a pure rust allocator (which must copy uninitialized memory, since it doesn't know what part of the buffer is initialized).

ecstatic-morse replied:
"valid" was an attempt to communicate "allocated memory whose access doesn't violate LLVM's pointer aliasing rules" without overwhelming the reader. We might want to move this stuff out (see next comment).

My only insight was that, according to the [reference](https://doc.rust-lang.org/reference/behavior-considered-undefined.html), undefined behavior includes "Reads of undef (uninitialized) memory".

I agree that my understanding was too strict, but I'm not sure what the proper constraint is.

Gankro commented on src/libcore/intrinsics.rs:
```diff
@@ -959,59 +959,131 @@ extern "rust-intrinsic" {
+    ///
+    /// * `dst` must be properly aligned.
+    ///
+    /// * The two regions of memory must *not* overlap.
```
This should be at the top, since all the other conditions are just "things that are true of any pointer accesses" (which suggests to me that they can potentially be omitted or centralized in the top-level module documentation and just linked?)

ecstatic-morse replied:
I think it would be a good idea to document the requirements for stores and loads at the top-level. My concern was that people coming to the docs from a search engine may not see the top-level documentation. There was also talk in #36450 of specifying alignment requirements for every function in `std::ptr`.

I'd like some clarification on what the precise requirements for loads and stores are regarding initializing memory, and what the precise definition of "valid" memory is.

alexcrichton commented:
@ecstatic-morse yeah seems good to me!

pietroalbini commented:
Ping from triage @steveklabnik! The author addressed the comments from the libs team.

steveklabnik commented:
@bors: r+ rollup

thanks!

bors commented:
:pushpin: Commit 827251e has been approved by `steveklabnik`

<!-- @bors r=steveklabnik 827251e92bef9bd613cf44e2dc074fa1dc71ea0f -->

Gankro commented:
My concerns weren't addressed, and I don't think this documentation should have landed as-is.

pietroalbini commented:
Ugh, why did GitHub mark the comments as resolved? `:/`

Gankro commented:
@steveklabnik could you revert this PR, since the author is also fine with not landing it?

ecstatic-morse commented:
@Gankro, @steveklabnik  I've pushed a new commit which adds a module-level definition of what a "valid" pointer is. The list of invariants for each function now links back to this definition.

I put the newly built docs in a [gh-pages branch on my fork](https://ecstatic-morse.github.io/rust/std/ptr/index.html) so that they can be more easily reviewed. I'll keep this up to date as I push changes.

I also removed the erroneous requirement that memory must be initialized if it is to be read from. Instead there is a [module-level discussion](https://ecstatic-morse.github.io/rust/std/ptr/index.html#uninitialized-memory) of how one cannot make any assumptions about the value of uninitialized memory. The example asserts that, even after calling `copy` on an uninitialized buffer, one cannot assume that the destination buffer is actually equal to the source. I think this effectively communicates the consequences of reading `undef` memory.

Also, do I need to open a new PR since this accidentally got merged and reverted?

Thanks!

rkruppe commented:
What one can and can't do with uninitialized memory is a bit up in the air, but it might very well be much more restrictive than what the current wording implies, so I'd prefer to err on the side of caution in these docs. I don't have specific wording to suggest but here are some things that one might get from the current text that might (likely IMO) not be true when we finally get Unsafe Code Guidelines on this matter:

- Uninitialized memory might change non-deterministically but after you load a value from it, it's some fixed bit pattern for which all the usual rules apply, e.g. after `let x = src[0];` expressions like `x + x == 2 * x` and `x >= 0` evaluate to true
- You can branch on its current value (e.g. `if src[0] == 0 { ... } else { ... }`) and it randomly picks one of the branches
- You can safely output uninitialized memory to a file or the network, it's just completely non-deterministic what bit pattern you get

Note that most of these aren't even true of LLVM's `undef`, and whatever we get in Rust may very well be much more restrictive than `undef` (e.g., one proposal is to allow only loading and storing, all arithmetic/comparisons/etc. would be instant UB).

steveklabnik commented:
> Also, do I need to open a new PR since this accidentally got merged and reverted?

Yes please!

ecstatic-morse commented:
@rkruppe I was under the impression that this already was rigorously defined behavior and that I was merely too dumb to find it :) Since defining these semantics is out of scope for this PR, I wonder if it's better to just not mention uninitialized memory at all. I'm assuming whatever choice gets made will permit loads and stores, which is all that is required by the functions in `ptr`. I'll leave out the module-level discussion, as well as omitting it from the list of invariants for each function.

ecstatic-morse commented:
[New PR](https://github.com/rust-lang/rust/pull/51016) opened, I'm closing this one.
